### ISSUE TITLE
import,backup: allow using `nodelocal` storage on a specific node

### ISSUE THREAD

**Issue by dt** [cockroachdb/cockroach#40543]
*TL;DR: make `nodelocal://2/foo.csv` work anywhere in the cluster, by allowing any node to ask node 2 for `foo.csv` via an RPC, similar to the approach in #17255.*

### Motivation

`nodelocal` storage has been a _constant_ source of confusion for our customers.

Despite our best efforts to push users to only use cloud storage, time and time again users are trying, and usually failing, to use local disks when trying get data in and out of cockroach. This is the most common usability complaint we get from IMPORT.

This has been a long-standing source of confusion and pain for users even before IMPORT, tracked in #19158 as well as other conversations, PoC notes, forum threads. Even customers with over a year of Cockroach operating experience still hit this trap regularly.

We've considered just taking `nodelocal` away to avoid this, and while some external cloud storage has obvious benefits, the overwhelming evidence is that for quick, one-off tasks like importing a file or backing up a small table for debugging, users want to reach for local disk instead of messing with cloud storage. Trying to convince them to do otherwise has proven to be a losing battle

The main driver of the problems with our current "nodelocal" is that it means "the local disk of the node evaluating this URI". In a distributed job however, that could be any one of many disks and this doesn't end up working how most users expect, and want it to work: They end up having no idea where their files are, or why they or cockroach can't find them. 

### Proposed Solution

Much of this could be fixed if `nodelocal` didn't mean "the local disk of the node evaluating it" but actually specified a particular disk, e.g. instead using `nodelocal://1/foo.csv` to specify`foo.csv` _on node 1's_ local disk. Such a URI does uniquely identify the same file, no matter which node in the cluster evaluates it, unlike the current schema.

Indeed, this scheme was suggested in #19158 and (and the code was changed to allow, but ignore, node IDs in URIs). At the time, the proposal was to return an error if a node other than the one identified in the URI was trying to read/write that file. This could help reduce confusion as the error could at least explain what was happening, e.g. "this nodelocal file URI specifies a file on node 1, but is being accessed on node 2.  Distributed job on multi-node clusters require cloud storage".

However, even better than an error might be to instead just make such a URI actually work.

If a user says `nodelocal://1/foo.csv` but node 2 gets scheduled to import that file, it could just ask node 1 for the content of `foo.csv` and carry with the job.  This is both clearer and easier for us and users, as it is clear what file on what node we're talking about, and it works from anywhere in the cluster when used in distributed jobs.

This is actually an approach we prototyped a long time ago in #17255 but didn't pursue at the time (it was a free-friday project and looked like it was going to make some code messier than I wanted to deal with at that time). 

### Future Work

In the future, this feature could pave the way for supporting more seamless importing of files on a client machine. Currently we require the user first put the file somewhere the cluster can fetch it (e.g. cloud storage, or manually pushing to every nodelocal path). With this feature, they would only need to manually push to one node. Even more exciting though is the potential to automatically push that file from a shell to its gateway, and then get a nodelocal URI back with the gateway's ID, and then be able to use that URI in jobs that any node in the cluster can reference. Load-balancers or firewalls may often mean that clients can't directly talk to every node in a cluster, so the ability to push to just their gateway and then have a URI that any other node can use is likely a requirement for any seamless client-file-import solution, so making `nodelocal` work across nodes looks like it could be a significant piece of that work.

**Comment by lvhui465021** [comment]
Has the nodelocal://nodeiD/file_name feature been implemented?

**Comment by dt** [comment]
#41990
